### DISCUSSION TITLE
Alarm:2 after toolchange

### DISCUSSION THREAD

**Discussion by Klippert** [terjeio/grblHAL#225] [Q&A] [answered]
Hello,

I've now had fairly long (1-3hour) programs running with 4-6 toolchanges, and it gennerally works really well.

Twice i've had it stop after a toolchange with alarm:2. 
I've had to reset the machine, home, reload, remove sections done, and restart, after which it runs the program smoothly.

It happens in the same situation both times, the tool is changed, TPW'd, and cycle is started and the machine moves up, to the first x,y coordinate, and then stops. 

At first I thought it was because the tool was long and the height i tried to lift to was outside the machines range, but now it also happended with a very short tool. 

Any suggestions on how to fix it would be most welcome.

Best regards
Kristoffer

**Comment by terjeio** [comment, marked as answer]
Alarm 2 is soft limits exceeded. If this is happening randomly then it could be hard to track down.

What is your $341 setting?
Is cycle start issued with sender button or button (switch) connected to the controller?

**Reply by Klippert** [reply]
I don't think it's entirely randomly, as it was in the same "sort of " place in both instances (albeit in two different programs).

I always use cycle start with the sender. (where you considering jitter from a possible start button?) 
Would that have been able to mess up the ysytem?

I'll look up my 341 = 2, so manual touch off @G53.9.

Have you got any suspicion ? 
Alternatively i can capture the console and .nc next time it happens, and we can try debugging it.

**Reply by terjeio** [reply]
If it is repeatable then it would be easier to track down. Did the spindle spin up before the alarm?

> I always use cycle start with the sender. (where you considering jitter from a possible start button?)

Not really, but you can never know.

> I'll look up my 341 = 2, so manual touch off @G53.9.
Have you got any suspicion ?

No, the exit routine is the same as for manual homing which a number of users use without any issues I know of. 

> Alternatively i can capture the console and .nc next time it happens, and we can try debugging it.

`$#` output and the machine position when the alarm happens would be interesting to see. 
If the error is repeatable with the same .nc file that would help too.

**Reply by Klippert** [reply]
okie. i'll update when it happens next. Then we will see if there is a  pattern to it. :-)
Thanks again for your help.

I did have an additional thought, and that is if there is a way to use a input on the breakout board to trigger the $tpw? - currently that's the only interaction that i can't do directly from the machine. 

/Kristoffer

**Reply by terjeio** [reply]
> I did have an additional thought, and that is if there is a way to use a input on the breakout board to trigger the $tpw?

Not yet - could be useful to be able to bind a button to a macro? On my todo list is adding a sowftware crossbar to some drivers where a number of pins can be bound to predefined pin functions. The Pi Pico driver is the first I want to experiment with in this regard.

**Reply by Klippert** [reply]
Alternatively using a controller input of some sort - like the digital input ST0-ST3 on phil's breakout bord for the Teensy (i don't know if such inputs also exist on other breakout boards/controllers?)

**Reply by terjeio** [reply]
> Alternatively using a controller input of some sort

Tht is what I was thinking about, to make it flexible bind the input (a button) to a macro. It is also possible to trigger execution of the $TP function call:

https://github.com/terjeio/grblHAL/blob/2d871b2d0a71193df8015f98da1bc565d12341d0/grbl/tool_change.c#L423-L426

>  like the digital input ST0-ST3 on phil's breakout bord for the Teensy (i don't know if such inputs also exist on other breakout boards/controllers?

Yes - but not all boards have free input pins, those that have should allow binding an input pin to a user selectable function or a macro.

**Reply by Klippert** [reply]
Hello - here we go the exact same error.

I got about half way through my programme, after third toolchange (to T12), I did the Probing with $TPW - click cycle start, and it goes to position above the workpiece and stops with the alarm.

To move on i do:
Reset
Home
edit the program and cut out everything from 
(Ramp1)
to
(2D Chamfer1)
hit save, close the editor,

Then  i Probe 
![image](https://user-images.githubusercontent.com/53910032/107053951-c53e0780-67cf-11eb-8f94-45cbb30b8eaa.png)
as you can see i can't establish a tool length offset. I actually have to close iosender, reopen it, reload the program. 
home again
Probe again - with everything set :)
![image](https://user-images.githubusercontent.com/53910032/107054700-c15eb500-67d0-11eb-88c0-2cee9b0bfec0.png)
and i hit cycle start, and things work out correctly from here.

Below is the console output:
GrblHAL 1.1f ['$' for help]
[MSG:'$H'|'$X' to unlock]
[MSG:Caution: Unlocked]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[G54:90.000,73.000,-76.625,0.000]
[G55:0.000,0.000,0.000,0.000]
[G56:0.000,0.000,0.000,0.000]
[G57:0.000,0.000,0.000,0.000]
[G58:0.000,0.000,0.000,0.000]
[G59:0.000,0.000,0.000,0.000]
[G59.1:0.000,0.000,0.000,0.000]
[G59.2:0.000,0.000,0.000,0.000]
[G59.3:1.000,519.155,0.000,0.000]
[G28:0.000,0.000,0.000,0.000]
[G30:0.000,0.000,0.000,0.000]
[G59.6:0.000,0.000,0.000,0.000]
[G92:0.000,0.000,0.000,0.000]
[HOME:0.000,0.000,0.000,0.000:0]
[TLO:0.000,0.000,0.000,0.000]
[PRB:0.000,0.000,0.000,0.000:0]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[VER:1.1f(IMXRT1062).20201103:]
[OPT:VNMZHSL+,35,1024,4,0]
[NEWOPT:*FLASH,ES,TC]
[DRIVER VERSION:201014]
[DRIVER OPTIONS:USB.2]
[BOARD:T41U5XBB]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[G54:90.000,73.000,-76.625,0.000]
[G55:0.000,0.000,0.000,0.000]
[G56:0.000,0.000,0.000,0.000]
[G57:0.000,0.000,0.000,0.000]
[G58:0.000,0.000,0.000,0.000]
[G59:0.000,0.000,0.000,0.000]
[G59.1:0.000,0.000,0.000,0.000]
[G59.2:0.000,0.000,0.000,0.000]
[G59.3:1.000,519.155,0.000,0.000]
[G28:0.000,0.000,0.000,0.000]
[G30:0.000,0.000,0.000,0.000]
[G59.6:0.000,0.000,0.000,0.000]
[G92:0.000,0.000,0.000,0.000]
[HOME:0.000,0.000,0.000,0.000:0]
[TLO:0.000,0.000,0.000,0.000]
[PRB:0.000,0.000,0.000,0.000:0]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[VER:1.1f(IMXRT1062).20201103:]
[OPT:VNMZHSL+,35,1024,4,0]
[NEWOPT:*FLASH,ES,TC]
[DRIVER VERSION:201014]
[DRIVER OPTIONS:USB.2]
[BOARD:T41U5XBB]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[G54:90.000,73.000,-76.625,0.000]
[G55:0.000,0.000,0.000,0.000]
[G56:0.000,0.000,0.000,0.000]
[G57:0.000,0.000,0.000,0.000]
[G58:0.000,0.000,0.000,0.000]
[G59:0.000,0.000,0.000,0.000]
[G59.1:0.000,0.000,0.000,0.000]
[G59.2:0.000,0.000,0.000,0.000]
[G59.3:1.000,519.155,0.000,0.000]
[G28:0.000,0.000,0.000,0.000]
[G30:0.000,0.000,0.000,0.000]
[G59.6:0.000,0.000,0.000,0.000]
[G92:0.000,0.000,0.000,0.000]
[HOME:0.000,0.000,0.000,0.000:0]
[TLO:0.000,0.000,0.000,0.000]
[PRB:0.000,0.000,0.000,0.000:0]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[G54:90.000,73.000,-76.625,0.000]
[G55:0.000,0.000,0.000,0.000]
[G56:0.000,0.000,0.000,0.000]
[G57:0.000,0.000,0.000,0.000]
[G58:0.000,0.000,0.000,0.000]
[G59:0.000,0.000,0.000,0.000]
[G59.1:0.000,0.000,0.000,0.000]
[G59.2:0.000,0.000,0.000,0.000]
[G59.3:1.000,519.155,0.000,0.000]
[G28:0.000,0.000,0.000,0.000]
[G30:0.000,0.000,0.000,0.000]
[G59.6:0.000,0.000,0.000,0.000]
[G92:0.000,0.000,0.000,0.000]
[HOME:0.000,0.000,0.000,0.000:7]
[TLO:0.000,0.000,0.000,0.000]
[PRB:0.000,0.000,0.000,0.000:0]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[PRB:1.000,519.155,-51.750,0.000:1]
[PRB:1.000,519.155,-51.740,0.000:1]
[GC:G0 G54 G17 G21 G91 G94 G49 G98 G50 M5 M9 T0 F25 S0.]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F25 S0.]
[PRB:1.000,519.155,-51.955,0.000:1]
[PRB:1.000,519.155,-51.950,0.000:1]
[MSG:Remove any touch plate and press cycle start to continue.]
[MSG:]
[PRB:1.000,519.155,-52.445,0.000:1]
[PRB:1.000,519.155,-52.440,0.000:1]
[MSG:Remove any touch plate and press cycle start to continue.]
[MSG:]
[PRB:1.000,519.155,-48.415,0.000:1]
[PRB:1.000,519.155,-48.405,0.000:1]
[MSG:Remove any touch plate and press cycle start to continue.]
ALARM:2
[MSG:Reset to continue]



And the program i can't seem to attach

**Reply by Klippert** [reply]
[Alu-clamp-v3.txt](https://github.com/terjeio/grblHAL/files/5933656/Alu-clamp-v3.txt)

**Reply by Klippert** [reply]
ok. after third last toolchange it does the same again. - so there must be something where it tries to go up outside the machine boundary. - but i don't know why... ?

**Reply by terjeio** [reply]
Your program is with the limits of my mini-mill so I can test it here. If it works for me I think you should try the test branch as this also reports the tool length offset reference on a `$#` request.

**Reply by Klippert** [reply]
![image](https://user-images.githubusercontent.com/53910032/107092428-df91d880-6803-11eb-8530-11ec6c6108f2.png)
when i use "safe retracts: Clearance height" it seems to work, and when i use G28 it seems to fail. 

Any clue as what that might mean?

**Reply by terjeio** [reply]
Any clue as what that might mean?

A bug? Rounding error?

Can you try with setting G28 a bit lower than absolute Z0? After homing issue:

 `G53G0Z-1`
`G28.1` 

to do that.

The tool change code retracts the spindle as well, to `Z-1` to avoid hitting the limit switch. The `G28` move prior to toolchange in your code is thus not neccesary. But it would still be nice to know why it fails.

BTW, I have run your program a few times without any errors. I did not change the tool though, I just probed with the same each time.

**Reply by Klippert** [reply]
I can certainly give that a shot. ok. same result. Alarm:2 - not at the first toolchange but after the second again. This time a threadmill which is 10mm longer than the two previous tools. so I do think we are on to something with it being Z that hits the limit.

**Reply by S2000Stefan** [reply]
I have also looked at @Klippert's problem on my machine. Unfortunately I have to 
confirm that I have exactly the same error, alarm 2, as Kristoffer.

The error does not occur with the IoSender Beta 8.7 from 17.01.
I can only reproduce the error with IoSender version 8.8 from 30.01., and the error also occurs with the new version from 31.01.
If I change from IoSender Beta 8.7 from 17.01 to Beta 8.8, the error is gone. Only after a restart of the 
controller, the error is present again with Beta 8.8. It is apparently important to retart the controller beforehand in order to reproduce the error.
I rule out rounding errors, as the Z height is no longer correct at all after the first tool change after the $TPW.
In my case, the MPos Z-1mm should be 117.904mm, but it is 16.694mm for Alarm 2. I did not change the tool, so the Z height should be identical.
![Z_richtig](https://user-images.githubusercontent.com/35189051/107146020-cee97b80-6945-11eb-8d09-726b5a34645a.jpg)
![Z_falsch](https://user-images.githubusercontent.com/35189051/107146027-d4df5c80-6945-11eb-82b4-d11cdafe6662.jpg)

So it is also understandable why IoSender goes to alarm 2, the next lines in the program code show the Z height at 23mm, which is outside the softlimits.

I also noticed that both LEDs in the main window light up after setting the reference point. Is this the way it should be?

I also noticed something else about Kristoffer's edition of the console:

> Below is the console output:
GrblHAL 1.1f ['$' for help]
[MSG:'$H'|'$X' to unlock]
[MSG:Caution: Unlocked]
[GC:G0 G54 G17 G21 G90 G94 G49 G98 G50 M5 M9 T0 F0 S0.]
[G54:90.000,73.000,-76.625,0.000]
[G55:0.000,0.000,0.000,0.000]
[G56:0.000,0.000,0.000,0.000]
[G57:0.000,0.000,0.000,0.000]
[G58:0.000,0.000,0.000,0.000]
[G59:0.000,0.000,0.000,0.000]
[G59.1:0.000,0.000,0.000,0.000]
[G59.2:0.000,0.000,0.000,0.000]
[G59.3:1.000,519.155,0.000,0.000]
[G28:0.000,0.000,0.000,0.000]
[G30:0.000,0.000,0.000,0.000]
[G59.6:0.000,0.000,0.000,0.000]   XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
[G92:0.000,0.000,0.000,0.000]
[HOME:0.000,0.000,0.000,0.000:0]
[TLO:0.000,0.000,0.000,0.000]
[PRB:0.000,0.000,0.000,0.000:0]

Is the line I have drawn with XXXX OK? G59.6?

**Reply by Klippert** [reply]
Thanks. I should add i'm using this version:
![image](https://user-images.githubusercontent.com/53910032/107146742-4de0b300-694a-11eb-8404-ae014c037166.png)

**Reply by terjeio** [reply]
> Is the line I have drawn with XXXX OK? G59.6?

No it is not, it is a minor bug in the version @Klippert is using.

@S2000Stefan Are you probing at g59.3 too when it fails, or on the workpiece?

There are siginficant changes to the code for the _Tool length offset_ tab from 8.7 to 8.8, it could well be that I have fixed broken bits and broken previously working bits of code. I'll look into that again.

**Reply by S2000Stefan** [reply]
> Are you probing at g59.3 too when it fails, or on the workpiece?

Yes, but only via the tool change function of the IoSender. "Manual touch off @ G59.3"
Z zero was on the workpiece surface. Touched with a macro.
There seems to be something wrong with the calculation of the TLO after the $TPW command.
The result comes directly after a tool change command and $TPW then comes alarm 2 when the machine moves back over the workpiece to position.
![Unbenannt2](https://user-images.githubusercontent.com/35189051/107151079-04e82900-6961-11eb-8e69-f3df76d69b4a.jpg)

As you can see, the Z height calculation is not correct at all; in this case, it should definitely be in the plus range in the DRO.

**Reply by terjeio** [reply]
The problem with the _Alu-clamp-v3_ program is the G28 move to the Z home position pre tool change. On cycle start after a tool change Z is moved to the home position minus 1mm, then to the original XY position and then _the controlled point_ (the tool tip) is moved to the original position. Since the original position is Z home due to the G28 move then any change that set a tool offset that is in the positive direction (a longer tool) is going to fail. This explains why this is true:

> when i use "safe retracts: Clearance height" it seems to work, and when i use G28 it seems to fail.

So moving the Z-axis to the home position before a tool change is not a good idea, and is not needed either as the tool change algorithm takes care of that. Backing off a little (by the "Clearance height") before the tool change is initiated is likely to be ok. 

_All_ calculations and moves related to tool changes are taking place in the controller using the machine coordinate system.
I therefore wonder if the difference between sender versions is just slight inaccuracies in the probing result when the tool is not changed? A tool length offset of one micrometer in the positive direction will be enough to trigger the alarm? Or am I missing something else?

**Reply by S2000Stefan** [reply]
> All calculations and moves related to tool changes are taking place in the controller using the machine coordinate system.
I therefore wonder if the difference between sender versions is just slight inaccuracies in the probing result when the tool is not changed? A tool length offset of one micrometer in the positive direction will be enough to trigger the alarm? Or am I missing something else?

Yes, your explanations make sense. However, I still believe that there is a bug in the IoSender from version 8.7.  ;-)
Here is a version from this morning. Same machine, same controller, exactly the same sequence until the tool change, cutter not reclamped or the like. The controller is also always restarted.
Therefore, after the $TPW command, approximately the same height should be entered in the z axis? See for yourself:

The pictures are once before the $TPW command and once after, both with IoSender 8.7 and IoSender 8.8.

Before:

![vor-tpw8 7](https://user-images.githubusercontent.com/35189051/107199605-f3516080-69f6-11eb-882b-0746ae133a71.jpg)
![vor-tpw8 8](https://user-images.githubusercontent.com/35189051/107199617-f64c5100-69f6-11eb-86ea-d876ef82e8f7.jpg)

After:

![nach-tpw8 7](https://user-images.githubusercontent.com/35189051/107199631-fb110500-69f6-11eb-9dc7-eca0e0697fb2.jpg)
![nach-tpw8 8](https://user-images.githubusercontent.com/35189051/107199659-fea48c00-69f6-11eb-8309-dd997e091fa1.jpg)

At Z level of -64.649, alarm2 always comes on. Since the Z height then goes into the soft limit according to the milling programme.
The process is always:
Homing of the machine.
Drive to G59.3 position with a macro.
Load the milling programme.
Execute Establish reference offset. If nothing else is ticked.
Set workpiece X Y zero point and Z zero with a macro.
Start programme.
Mill moves to G59.3 change position, I jog down, then execute a $TPW command.

With IoSender 8.7 the programme always runs completely, with IoSender 8.8 after the above procedure it always goes into Alarm 2.

**Reply by terjeio** [reply]
I have made a mistake in that the tool offset is sometimes set when the reference offset is set (both LEDs light up). I believe I have a fix and maybe I should upload that for others to test as well.

**Reply by S2000Stefan** [reply]
@terjeio i read [here](https://github.com/terjeio/grblHAL/discussions/255) this morning that you have uploaded a new [Edge version ](http://www.io-engineering.com/downloads) and that you have addressed the TLO problem from here.
After short tests this morning and some tool changes I can report that the TLO is now set correctly, but hopefully I will be able to do more intensive tests today at noon, if there are still problems I will report them.

Thanks again for the quick support.

I have also tested the milling programme from @Klippert here and it now runs through without any problems.
I would advise you to set G28 Kristoffer if you want to use it. Or at least clear the end position and then set G28 to have some buffer.

**Reply by Klippert** [reply]
Brilliant! Thanks @S2000Stefan - you're quick ;)

Just a question - what is the actual difference between using say "cleanance height" and G28? - As i understand it, both are setting a certain Z height that the tool will go to before moving x or y, to prevent crashes. Is that correct?

**Reply by S2000Stefan** [reply]
I wouldn't say that.  ;-)
 G28 and G30 are fixed points. You can also place Z0 at G28 on the machine bed, for example.
The best thing to do is to look [here](http://linuxcnc.org/docs/2.5/html/gcode/gcode.html#sec:G28-G28_1), where all the G-code commands are explained quite well.
I always look there when I don't understand a G-code.
I use G28, for example, after work to move the machine to a certain parking position and then to make a homing from this position in the morning.

**Reply by Klippert** [reply]
Cool. That's a nice list you posted there. I'm gonna keep that link handy. ;)

Ok, so let's say i use G28 as a safe  parking position - how does that tie in with the GRBL-Hal Parking function? 
Doesn't that park the machine at a specific position after each job?

**Reply by S2000Stefan** [reply]
> how does that tie in with the GRBL-Hal Parking function?

What do you mean by grblHal parking function?
Do you mean the safty door?
Or what do you mean?

As I understand it, you use G28 to bring your tool to maximum height when changing tools (G28 G91 Z0). It is not necessary to specify this in the PP as grblHal does this for you automatically.
The only thing is at the end of your milling programme, when you move your machine into a corner, grblHal does not do this automatically. First on Zmax then on X0 Y0 which you have stored in G28.
(G28 G91 Z0)
G90
(G28 G91 X0 Y0)
G90

**Reply by Klippert** [reply]
This setting:
![image](https://user-images.githubusercontent.com/53910032/109355925-442ed900-7880-11eb-9363-9861466e4e10.png)

the parking cycle and parking axis. 

And thank you very much for explaining G28 so elaborately. I will now set it and use it :-)

**Reply by S2000Stefan** [reply]
I think the settings refer to the safty door. In any case, I use this setting for it. 
I have a hardware switch connected to the safty door pin, which allows me to interrupt the work of the machine at any time so that I can safely make changes and then continue the milling job.

![SaftyDoor](https://user-images.githubusercontent.com/35189051/109378580-052a7300-78d4-11eb-81b0-affc3dad3ffc.jpg)

**Reply by Klippert** [reply]
Ah - thanks - now i understand how that works. - that is super helpful! but shouldn't it be called something other than Safety door? It's more like a pause function? 
I understand that it comes from opening the safetydoor to a big machining center, but i guess you would want a way to stop/pause things, so you don't have coolant and chips comming at you in order to pause things. :-) 

Anyhow, thanks for the brilliant description. I guess these forums by now contains most of a manual for GRBL-Hal  ;-)

**Reply by einencool** [reply]
The Pause function only stops moving, the „safety Door“ stops the Spindle and drives the Endmill out of the Workpiece. 
So both functions have their usecase :-)

**Reply by Klippert** [reply]
Is there a documentation of the parameters 56-59? - i'm a bit confused on the pull-out distance and the parking target.

definitely a usefull feature :)

**Reply by S2000Stefan** [reply]
$56 Parking pull-out distance - the distance the cutter travels at the speed set in $57 Parking pull-out rate after the safty door is triggered.
The spindle is then switched off and moves at the speed from $59, Parking fast rate, to the set height from $58 Parking target (note: machine position).
After pressing start, everything takes place in reverse order and the milling job continues.

**Reply by Klippert** [reply]
ah brilliant. is there a grbl-hal manual somewhere? or should we start making one?

**Reply by terjeio** [reply]
> is there a grbl-hal manual somewhere?

There are some information in the [grblHAL Wiki](https://github.com/terjeio/grblHAL/wiki), mainly supplements to the [legacy Grbl Wiki](https://github.com/gnea/grbl/wiki). The legacy wiki should be pulled in?
